pkg/controller/elasticsearch/driver/upscale.go

```diff
@@ -11,50 +11,116 @@ import (
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/nodespec"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/observer"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/settings"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/version/zen1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/version/zen2"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 )
 
+type upscaleCtx struct {
+	k8sClient           k8s.Client
+	es                  v1alpha1.Elasticsearch
+	scheme              *runtime.Scheme
+	observedState       observer.State
+	esState             ESState
+	upscaleStateBuilder *upscaleStateBuilder
+}
+
 // HandleUpscaleAndSpecChanges reconciles expected NodeSpec resources.
 // It does:
 // - create any new StatefulSets
 // - update existing StatefulSets specification, to be used for future pods rotation
 // - upscale StatefulSet for which we expect more replicas
+// - limit master node creation to one at a time
 // It does not:
 // - perform any StatefulSet downscale (left for downscale phase)
 // - perform any pod upgrade (left for rolling upgrade phase)
 func HandleUpscaleAndSpecChanges(
-	k8sClient k8s.Client,
-	es v1alpha1.Elasticsearch,
-	scheme *runtime.Scheme,
-	expectedResources nodespec.ResourcesList,
+	ctx upscaleCtx,
 	actualStatefulSets sset.StatefulSetList,
+	expectedResources nodespec.ResourcesList,
 ) error {
-	// TODO: there is a split brain possibility here if going from 1 to 3 masters or 3 to 7.
-	//  We should add only one master node at a time for safety.
-	//  See https://github.com/elastic/cloud-on-k8s/issues/1281.
-
-	for _, nodeSpecRes := range expectedResources {
-		// always reconcile config (will apply to new & recreated pods)
-		if err := settings.ReconcileConfig(k8sClient, es, nodeSpecRes.StatefulSet.Name, nodeSpecRes.Config); err != nil {
+	// adjust expected replicas to control nodes creation and deletion
+	adjusted, err := adjustResources(ctx, actualStatefulSets, expectedResources)
+	if err != nil {
+		return err
+	}
+	// reconcile all resources
+	for _, res := range adjusted {
+		if err := settings.ReconcileConfig(ctx.k8sClient, ctx.es, res.StatefulSet.Name, res.Config); err != nil {
 			return err
 		}
-		if _, err := common.ReconcileService(k8sClient, scheme, &nodeSpecRes.HeadlessService, &es); err != nil {
+		if _, err := common.ReconcileService(ctx.k8sClient, ctx.scheme, &res.HeadlessService, &ctx.es); err != nil {
 			return err
 		}
-		ssetToApply := *nodeSpecRes.StatefulSet.DeepCopy()
-		actual, alreadyExists := actualStatefulSets.GetByName(ssetToApply.Name)
-		if alreadyExists {
-			ssetToApply = adaptForExistingStatefulSet(actual, ssetToApply)
+		if err := sset.ReconcileStatefulSet(ctx.k8sClient, ctx.scheme, ctx.es, res.StatefulSet); err != nil {
```

Comment:
Should we not update the expectation for this statefulset if it is updated ?
e.g.:
1. an upscale from 2 to 3 masters has just be done
2. in the meantime a new statefulset with some new masters has been added by the user
3. if we have a stale cache for the first statefulset we would add a fourth master with a m_m_n of 2

Reply:
My above comment assumes that the new `statefulset` is at the top of the list returned by `adjustResources` and would have precedence over the first one.

Reply:
I think you are correct. Also maybe I am missing something but it looks to me like the limiting is only happening on replica increases of existing ssets but does not limit the addition of completely new ones

Reply:
Yes so I just confirmed that my reading of the code is correct: 
If you go from 
```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1alpha1
kind: Elasticsearch
metadata:
  name: elasticsearch-test
spec:
  version: 7.3.0
  nodes:
  - name: master-1
    config:
      node.master: true
      node.data: true
    nodeCount: 3
```
to 
```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1alpha1
kind: Elasticsearch
metadata:
  name: elasticsearch-test
spec:
  version: 7.3.0
  nodes:
  - name: master-1
    config:
      node.master: true
      node.data: true
    nodeCount: 3
  - name: master-2
    config:
      node.master: true
      node.data: true
    nodeCount: 3
```

The new master nodes in the `master-2` sset will be created all at once.